docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+
+```bash
+/var/www/SS_acme/ > upgrade-code environment --write
 ```
 
-This will ensure that your `.htaccess` and `index.php` are set to a reasonable default value for a clean installation.
+If your `_ss_environment.php` file contains unusual logic (conditional statements or loops), you will get a warning. `upgrade-code` will still try to convert the file, but you should double-check the output.
 
-### Renamed and namespaced classes
+Omit the `--write` flag to do a dry-run.
 
-Nearly all core PHP classes in SilverStripe have been namespaced. For example, `DataObject` is now called `SilverStripe\ORM\DataObject`.
+### Manually convert `_ss_environment.php` to `.env`
 
-For a full list of renamed classes, check the `.upgrade.yml` definitions in each module.
+Create a `.env` file in the root of your project. Replace `define` statements from `_ss_environment.php` with `KEY=VALUE` pairs in `.env`.
 
-After installing, run the upgrader upgrade command:
+Most SilverStripe 3 environment variables have been carried over to SilverStripe 4. See [Environment Management docs](/getting_started/environment_management/) for the full list of available variables. Your `.env` file can contain environment variables specific to your project as well.
+
+The global array `$_FILE_TO_URL_MAPPING` has been removed and replaced with the `SS_BASE_URL` environment variable. `SS_BASE_URL` expects an absolute url with an optional protocol. The following are values would be valid entries for `SS_BASE_URL`:
+* `http://localhost/`
+* `https://localhost/`
+* `//localhost/`                                                                                                        
+
+For example, take the following `_ss_environment.php` file.
+```php
+<?php
+// Environment
+define('SS_ENVIRONMENT_TYPE', 'dev');
+define('SS_DEFAULT_ADMIN_USERNAME', 'admin');
+define('SS_DEFAULT_ADMIN_PASSWORD', 'password');
+$_FILE_TO_URL_MAPPING[__DIR__] = 'http://localhost';
+
+// Database
+define('SS_DATABASE_CHOOSE_NAME', true);
+define('SS_DATABASE_CLASS', 'MySQLDatabase');
+define('SS_DATABASE_USERNAME', 'root');
+define('SS_DATABASE_PASSWORD', '');
+define('SS_DATABASE_SERVER', '127.0.0.1');
 ```
-cd ~/my-project-root
-~/.composer/vendor/bin/upgrade-code upgrade ./app/src --write
+
+The equivalent `.env` file will look like this.
+```bash
+## Environment
+SS_ENVIRONMENT_TYPE="dev"
+SS_DEFAULT_ADMIN_USERNAME="admin"
+SS_DEFAULT_ADMIN_PASSWORD="password"
+SS_BASE_URL="http://localhost/"
+
+## Database
+SS_DATABASE_CHOOSE_NAME="true"
+SS_DATABASE_CLASS="MySQLDatabase"
+SS_DATABASE_USERNAME="root"
+SS_DATABASE_PASSWORD=""
+SS_DATABASE_SERVER="127.0.0.1"
 ```
 
-### Switching to new `app/src` structure
+### Cleaning up `mysite/_config.php` after your environment configuration upgrade
+
+You'll need to clean up your `mysite/_config.php` file after upgrading your environment file.
```

Comment:
Might be worth mentioning that any `Config->update()` calls will slow down your entire execution in the way the config system is implemented in 4.x (or rather speed it up if you don't use it) - as a motivator for fixing up this file.

Reply:
Not sure I'm familiar with this. Are config value cached?